### ISSUE TITLE
page layout ignored when generating pdf from template pdf

### ISSUE THREAD

**Issue by ckhall** [prawnpdf/prawn#293]
this happens in both prawn 0.12.0 and 1.0.0rc1

code sample to show issue:

```
# generate a landscaped pdf page and save it
Prawn::Document.generate 'landscape.pdf', :page_layout => :landscape do
  font 'Helvetica'
  text 'hello world'
  start_new_page # generates new page with proper page layout
  text 'page still in landscape'
end

Prawn::Document.generate 'test1.pdf', :template => 'landscape.pdf' do
  font 'Helvetica'
  go_to_page(page_count)
  start_new_page # generates page, but ignores layout of template
  text 'adding page to template not using :page_layout'
end

Prawn::Document.generate 'test2.pdf', :template => 'landscape.pdf', :page_layout => :landscape do
  font 'Helvetica'
  go_to_page(page_count)
  start_new_page # generates page, but ignores layout of template and :page_layout option
  text 'adding page to template and using :page_layout => :landscape in #generate'
  start_new_page(:page_layout => :landscape)
  text 'adding another page and using :page_layout => :landscape in #start_new_page'
end
```

**Comment by AndrewSwerlick** [comment]
Obviously this issue is pretty much dead given templates is no longer supported, but for those stuck on an older release using the templates code, I was able to figure out the cause of this bug and a workaround. For others who stumble across this bug while googling, I solved it by monkey patch PDF::Core::Page like this

    module PDF
      module Core
        class Page
          def size
            return @size if defined?(@size) && @size

            case(layout)
            when :portrait
               dimensions[2,2]
            when :landscape
               dimensions[2,2].reverse
            end
          end
        end
      end
    end

Basically the problem is that when templates are used, the size instance variable does not get set, so call to page.size just return the dimensions. However, when this value is fed back into Page.new to create a new page, it reverses the coordinates if the layout is landscape. Since page.size already gives you the coordinates in landscape, reversing them makes it portrait.

**Comment by practicingruby** [comment]
@AndrewSwerlick There is an effort going on at https://github.com/prawnpdf/prawn-templates to get templates working again as an extension. Please post this over there, and @bvogel will check it out.

**Comment by AndrewSwerlick** [comment]
@practicingruby Thanks. I was also wondering if I should post something in pdf-core. Arguably this is a pdf core bug, since currently if you ever pass :object_id to Page.new, it will fail to set the size instance variable, and as a result, incorrectly report the size for any landscape based layouts. That being said, I don't know if any code path that doesn't use templates would encounter this bug, because I don't know if anything else uses the :object_id option

**Comment by practicingruby** [comment]
I think @bvogel can help decide that too. I had removed some (but not all) template-specific functionality from PDF::Core in a somewhat sloppy fashion, and so there may be more that needs to be extracted.

**Comment by bvogel** [comment]
Hi guys, 
I wouldn't reopen the issue here. As Gregory pointed out, there is no way for this issue to occur outside the realm of templates. My understanding is that prawn-templates should know the page layout and set it accordingly.

**Comment by AndrewSwerlick** [comment]
@bvogel I wasn't necessarily proposing to reopen the item, I just wanted to leave the workaround there for posterities sake. That being said, I think this may be outside of the realm of templates, since it's really tied to the behavior of `PDF::Core::Page`, so long as `PDF::Core::Page` allows you to pass an `:object_id` to the initialize options hash. When you build a page from an object with a landscape orientation, `PDF::Core::Page` will never properly report the size, and this will always screw up how `Prawn::Document#start_new_page` behaves. If the only reason `PDF::Core::Page` accepts `:object_id` is for templates, then probably that should be removed from the core code. If there are other reasons, then I think this is a broader `PDF::Core` bug.

**Comment by bvogel** [comment]
Andrew, you are correct about the code removal, template is the only use case to use object_ids as base for page objects.
